src/rpc/mining.cpp

```diff
@@ -610,29 +636,32 @@ UniValue getblocktemplate(const UniValue& params, bool fHelp)
         (mempool.GetTransactionsUpdated() != nTransactionsUpdatedLast && GetTime() - nStart > 5))
     {
         // Clear pindexPrev so future calls make a new block, despite any failures from here on
-        pindexPrev = NULL;
+        pindexPrev = nullptr;
 
-        // Store the pindexBest used before CreateNewBlockWithKey, to avoid races
         nTransactionsUpdatedLast = mempool.GetTransactionsUpdated();
+
+        // If we're going to use the precomputed coinbase (empty block) and there are
+        // transactions waiting in the mempool, make sure that on the next call to this
+        // RPC, we consider the transaction count to have changed so we return a new
+        // template (that includes these transactions) so they don't get stuck.
+        if (next_coinbase_mtx && mempool.mapTx.size() > 0) nTransactionsUpdatedLast = 0;
```

Comment:
I agree this solves the documented edge case, because this variable is fed into the `longpollid` returned to the caller, and is also the fallback static if the caller does not provide `longpollid`.

This should use the API that correctly takes the mempool lock:
```suggestion
        if (next_coinbase_mtx && mempool.size() > 0) nTransactionsUpdatedLast = 0;
```